malfeasance2/publisher.go

```diff
@@ -0,0 +1,196 @@
+package malfeasance2
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	"golang.org/x/exp/maps"
+
+	"github.com/spacemeshos/go-spacemesh/codec"
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/p2p/pubsub"
+	"github.com/spacemeshos/go-spacemesh/sql"
+	"github.com/spacemeshos/go-spacemesh/sql/atxs"
+	"github.com/spacemeshos/go-spacemesh/sql/malfeasance"
+	"github.com/spacemeshos/go-spacemesh/sql/marriage"
+)
+
+type Publisher struct {
+	logger    *zap.Logger
+	db        sql.Executor
+	sync      syncer
+	tortoise  tortoise
+	publisher pubsub.Publisher
+}
+
+func NewPublisher(
+	logger *zap.Logger,
+	db sql.Executor,
+	sync syncer,
+	tortoise tortoise,
+	publisher pubsub.Publisher,
+) *Publisher {
+	return &Publisher{
+		logger:    logger,
+		db:        db,
+		sync:      sync,
+		tortoise:  tortoise,
+		publisher: publisher,
+	}
+}
+
+func (p *Publisher) PublishATXProof(ctx context.Context, nodeID types.NodeID, proof []byte) error {
+	marriageID, err := marriage.FindIDByNodeID(p.db, nodeID)
+	switch {
+	case errors.Is(err, sql.ErrNotFound): // smesher is not married
+		malicious, err := malfeasance.IsMalicious(p.db, nodeID)
+		if err != nil {
+			return fmt.Errorf("check if smesher is malicious: %w", err)
+		}
+		if malicious {
+			p.logger.Debug("smesher is already marked as malicious", zap.String("smesher_id", nodeID.ShortString()))
+			return nil
+		}
+		if err := malfeasance.AddProof(p.db, nodeID, nil, proof, int(InvalidActivation), time.Now()); err != nil {
+			return fmt.Errorf("setting malfeasance proof: %w", err)
+		}
+		atxID, err := atxs.GetFirstIDByNodeID(p.db, nodeID)
+		if err != nil {
+			return fmt.Errorf("getting atx id: %w", err)
+		}
+		p.tortoise.OnMalfeasance(nodeID)
+		return p.publish(ctx, []types.NodeID{nodeID}, []types.ATXID{atxID}, proof, InvalidActivation)
+	case err != nil:
+		return fmt.Errorf("getting equivocation set: %w", err)
+	default: // smesher is married
+	}
+
+	// Combine IDs from the present equivocation set for atx.SmesherID and IDs in atx.Marriages.
+	set, err := marriage.NodeIDsByID(p.db, marriageID)
```

Comment:
Would it make sense to have `EquivocationSetByNodeID` function in `sql/marriage` package which returns the equivocation set by means of an SQL join to reduce number of queries? Granted, hopefully we will not ever see a lot of malfeasance proofs being published at the same time (except maybe due to some kind of sophisticated attack), but as I see `marriage.FindIDByNodeID` is used together with `marriage.NodeIDsByID` a few times in the code

Reply:
`NodeIDsByID` is this query:

```sql
SELECT pubkey FROM marriages WHERE id = $1
```

To get the `marriageID` from a nodeID calling `marriage.FindIDByNodeID` before is needed. That function does this query:

```sql
SELECT id FROM marriages WHERE pubkey = $1
```

Sure one could combine these two into a single query like

```sql
SELECT pubkey
FROM marriages
WHERE id = (SELECT id FROM marriages WHERE pubkey = $1)
```

but I don't quite see how this improves the situation? Especially since except from here we often also need the `marriageID` for other things than just looking up the other `nodeIDs` as well?